### ISSUE TITLE
Ldap cannot bind users correctly

### ISSUE THREAD

**Issue by n-borges** [snipe/snipe-it#13564]
### Debug mode

- [X] I have enabled debug mode
- [X] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

### Describe the bug

Hi everyone

I'm trying to configure a Snipe-IT instance to work with a Zentyal LDAP server.
I have configured several other services to work with our LDAP, and they seem to work well; Snipe-IT, though, is giving me some errors I cannot quite understand.

Other services i have setup seem to authenticate users with the following:
`cn=<first name> <last name>,dc=example,dc=com`

While Snipe-IT, no matter how I try to tweak the configuration, sends the following to my LDAP:
`samaccountname=<whatever is entered in the username field>,dc=example,dc=com`

In the configuration page, i can successfully test "LDAP Syncronisation". All the expected users show up, according to the filters i have configured.
"Test LDAP Login", though, fails with: `Login Failed. <user> did not successfully bind to LDAP.`

I tried every configuration that came to my mind to try and debug this, but nothing seems to work. 
Do you have any suggestions? I'll put a screenshot of my configs here

![ldap_configs](https://github.com/snipe/snipe-it/assets/81313286/a936efa8-e8c9-46cd-a360-d60600253e70)

Thanks for your help!

### Reproduction steps

1. deploy a Snipe-IT instance
2. deploy a Zentyal instance
3. try to configure LDAP

### Expected behavior

Snipe-IT should authenticate through LDAP successfully

### Snipe-IT Version

6.1.2

### Operating System

Alpine (Docker)

### Web Server

Apache

### PHP Version

8.1.22

**Comment by Alpha6333** [comment]
Same Problem.

**Comment by CuH-Sys** [comment]
Can confirm problem with docker installation.

**Comment by snipe** [comment]
@uberbrady can you take a look at this when you get a moment?

**Comment by uberbrady** [comment]
I’d be surprised if this’d work, but can you try LDAP Auth query of `cn=` ?

Snipe-IT tends to expect the thing you log in as to be the ‘username’ and I’m not sure if we’re smart enough to let you authenticate with something else entirely, but then pluck the ‘username’ out of the LDAP results and correctly use that. So that’s why I’m pretty skeptical that it’d work.

**Comment by n-borges** [comment]
> I’d be surprised if this’d work, but can you try LDAP Auth query of `cn=` ?
> 
> Snipe-IT tends to expect the thing you log in as to be the ‘username’ and I’m not sure if we’re smart enough to let you authenticate with something else entirely, but then pluck the ‘username’ out of the LDAP results and correctly use that. So that’s why I’m pretty skeptical that it’d work.

Unfortunately it doesn't work; I think the query is still incomplete as no OUs are specified. I guess the solution, as you suggested, would be to decouple what snipeit considers as its "username" and what fields are used to actually authenticate against the LDAP server. 

I suppose that other services do exactly this: query the server using the provided username, get the object, and then build the authentication query using the retrieved data. 

I cannot quite understand if we fell in some obscure usecase or if this is a common problem. Do you think a fix for this would be too much of a trouble to implement?

**Comment by uberbrady** [comment]
It's relatively obscure, but I've heard people ask for something similar to this before, so I wouldn't say unique. The challenge here would be for how you go about storing the 'authentication string' separately from the 'username' for when you want to do fallthrough-authentication (for when an LDAP server is down). Or, maybe, we just allow it to 'not work' and leave it at that. I'm not sure. I think we have a PR up that _might_ handle this use case? This one: https://github.com/snipe/snipe-it/pull/11715 - if you have the ability to pull that down and test it, it'd be interesting if it managed to handle this use case - but flipping through the code I suspect it might not. But, if you can, do check it out and report back.

**Comment by n-borges** [comment]
I'm sorry for the late reply, but unfortunately this PR's version does not solve things either.

Do you think this is worth fixing/working on?

**Comment by DepaMarco** [comment]
Hi. Thanks for opening this bug report.
Same problem for me, version "v6.2.0 - build 11566 (master)"
Thanks.

**Comment by Alpha6333** [comment]
It is still not possible to use LDAP login.
I assume that I have set the data correctly.
Reading the LDAP table also works. Every time I try to test a user it does not work.

In the database, all users that were imported have the same hash in the PW entry.

**Comment by Alpha6333** [comment]
Hi all,

what is the status from this ticket.
Unfortunately it still does not work with the latest update that imported LDAP users can log in.

**Comment by Alpha6333** [comment]
Version v6.1.2 - build 10938 (master)

**Comment by Alpha6333** [comment]
Interessting result on cli

```
/var/www/html/snipe.xxxxx.com# sudo -u www-data php artisan ldap:troubleshoot

 WARNING: This command will make several attempts to connect to your LDAP server. Are you sure this is ok? (yes/no) [no]:
 > yes

STAGE 1: Checking settings
Determined LDAP hostname to be: ldap.xxxxxx.com
Performing DNS lookup of: ldap.xxxxxxx.com

In LdapTroubleshooter.php line 214:
                                                        
  dns_get_record(): A temporary server error occurred.  

```

**Comment by panzli** [comment]
I think this issue could be related to my issue, the login part of the  problem looks like the same from this issue : https://github.com/snipe/snipe-it/issues/13845

**Comment by DepaMarco** [comment]
@Alpha6333 I see you fixed LDAP Sync in https://github.com/snipe/snipe-it/issues/13845 (many thanks!), now I can login using "cn=xxx" (I'm in v6.3.0).
But how is "LDAP Authentication query" really used? I mean, in my case the BIND operation could only happens with "cn=" but the username I want to use is the field "uid=" (RDN identifier is "cn" in my "ou=People").
It seems that any value I put in that field is completely ignored.

**Comment by ProxayFox** [comment]
Hello, how are you going.

TLDR, I might have found a fix. Some DNS servers will block `DNS_ANY` requests

I've been play around with ldap and ran the same error as [Alpha6333](https://github.com/snipe/snipe-it/issues/13564#issuecomment-1798056180)

When I tried to fix it I found that there's apparently a [PHP bug](https://bugs.php.net/bug.php?id=73149) from what I can understand. Although my understanding is limited.

I was sure I'd gotten this working on another machine, so I went between the two and played with 
```php
<?php
$url = "ldap.google.com";
$result = dns_get_record($url);
print_r($result);
```

On the clean system where I'm getting the error, i get 
```txt
PHP Warning:  dns_get_record(): A temporary server error occurred. in /var/www/snipt_it_ecojoule/public/test.php on line 3
```

On my main system I get 
```text
Array
(
    [0] => Array
        (
            [host] => ldap.google.com
            [class] => IN
            [ttl] => 85431
            [type] => A
            [ip] => 216.239.32.58
        )

    [1] => Array
        (
            [host] => ldap.google.com
            [class] => IN
            [ttl] => 85431
            [type] => AAAA
            [ipv6] => 2001:4860:4802:32::3a
        )

)
```
I thought it might have been a missing package but it wan't a package
I then checked the fpm files for php.ini, php-fpm.conf, www,conf. but there was no notable difference
I can query with Dig and that works, Buuuuuut then I found [this](https://docs.mdoq.io/hc/en-gb/articles/13322420076945-dns-get-record-not-working), where it talked about how some DNS servers will block `DNS_ANY` requests. annnnnnd that started to make some sense.

The clean server is on Digitalocean and the working server is at home on a residential address.
When running `dig ldap.google.com any` on the Digitalocean box it doesn't return anything but when running that on the home server it returns addresses.
So it looks like in this example google is blocking Digitalocean droplets from running `DNS_ANY`

Soooooooo I was able to fix this by editing LdapTroubleshooter.php lines 213-215 and it worked, but now I have a new issue to resolve ahaha
```php
$this->info("Performing DNS lookup of: ".$parsed['
$ips = dns_get_record($parsed['host']);
$raw_ips = [];
```
with
```php
$this->info("Performing DNS lookup of: ".$parsed['
$ips = dns_get_record($parsed['host'], DNS_A + DNS_CNAME);
$raw_ips = [];
```

**Comment by DepaMarco** [comment]
Just upgraded my installation of Snipe-IT to version 8.0.4 but I'm still having the same issue...
It seems "LDAP Authentication query" is not used at all. Why forcing "uid=..." in the binding query to LDAP? It's a common used option to have "cn=" as DN and not the "uid=".
Please have a look, I've also posted here: [on Apr 16, 2025](https://github.com/grokability/snipe-it/issues/7067#issuecomment-2809038729)

**Comment by snipe** [comment]
@uberbrady Any thoughts on this?

**Comment by uberbrady** [comment]
So I'm looking at this method: https://github.com/grokability/snipe-it/blob/master/app/Models/Ldap.php#L92

and it looks like here:

https://github.com/grokability/snipe-it/blob/master/app/Models/Ldap.php#L98

we're assembling the `$userDn` by using the `ldap_username_field`, adding an `=`, putting in the username, then appending a `,` and the value of the BaseDN.


Here is where we assemble together a string with the `ldap_auth_filter_query` - 

https://github.com/grokability/snipe-it/blob/master/app/Models/Ldap.php#L116

We use take that as-is, and just appending the username to it - but we only use that here:

https://github.com/grokability/snipe-it/blob/master/app/Models/Ldap.php#L143

The actual *bind* happens before that - and that's with the `$userDn` value:

https://github.com/grokability/snipe-it/blob/master/app/Models/Ldap.php#L122

So the thought I have had (for a while) was to change the order of that - *first*, do the LDAP query - fetching all of the stuff about the user - and _then_ binding to the returned DN.

We've had a few folks try and offer PR's with changes like those up to us, but we always chicken out because it's LDAP, and it's Authentication, and it's scary.

**Comment by uberbrady** [comment]
Actually - it's a little harder than that - because we need to do that initial query as _someone_ - so we would probably have to do that as the Admin user. So that changes around the order of operation a little bit.

Hence us not having done it yet.
